pkg/runtime/serializer/codec_factory.go

```diff
@@ -308,3 +310,52 @@ func (f CodecFactory) SerializerForFileExtension(extension string) (runtime.Seri
 	}
 	return nil, false
 }
+
+type DirectCodecFactory struct {
+	CodecFactory
+}
+
+func (f DirectCodecFactory) EncoderForVersion(serializer runtime.Encoder, gv unversioned.GroupVersion) runtime.Encoder {
+	return DirectCodec{
+		struct {
+			runtime.Encoder
+			runtime.Decoder
+		}{
+			serializer,
+			nil,
+		},
+		f.CodecFactory.scheme,
+	}
+}
+
+func (f DirectCodecFactory) DecoderToVersion(serializer runtime.Decoder, gv unversioned.GroupVersion) runtime.Decoder {
+	return DirectCodec{
+		struct {
+			runtime.Encoder
+			runtime.Decoder
+		}{
+			nil,
+			serializer,
+		},
+		nil,
+	}
+}
+
+type DirectCodec struct {
+	runtime.Serializer
+	runtime.ObjectTyper
+}
+
+func (c DirectCodec) EncodeToStream(obj runtime.Object, stream io.Writer, overrides ...unversioned.GroupVersion) error {
+	gvk, err := c.ObjectTyper.ObjectKind(obj)
+	if err != nil {
+		return err
+	}
+	kind := obj.GetObjectKind()
+	kind.SetGroupVersionKind(gvk)
+	return c.Serializer.EncodeToStream(obj, stream, overrides...)
+}
+
+func (c DirectCodec) Decode(data []byte, defaults *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
+	return c.Serializer.Decode(data, defaults, into)
```

Comment:
You may want to clear the kind after you decode, to keep consistency with Codec.